storage/innobase/row/row0import.cc

```diff
@@ -3075,6 +3178,151 @@ static size_t get_buf_size()
       ;
 }
 
+/** Add fts index to the table
+@param table fts index to be added on the table */
+static void add_fts_index(dict_table_t *table)
+{
+  dict_index_t *fts_index= dict_mem_index_create(
+    table, FTS_DOC_ID_INDEX_NAME, DICT_UNIQUE, 2);
+  fts_index->page= FIL_NULL;
+  fts_index->cached= 1;
+  fts_index->n_uniq= 1;
+  /* Add fields for FTS_DOC_ID_INDEX */
+  dict_index_add_col(
+    fts_index, table,
+    &table->cols[table->n_cols - (DATA_N_SYS_COLS + 1)], 0);
+  dict_index_t *clust_index= UT_LIST_GET_FIRST(table->indexes);
+  for (ulint i= 0; i < clust_index->n_uniq; i++)
+    dict_index_add_col(fts_index, table, clust_index->fields[i].col,
+                       clust_index->fields[i].prefix_len);
+#ifdef BTR_CUR_HASH_ADAPT
+  fts_index->search_info= btr_search_info_create(fts_index->heap);
+  fts_index->search_info->ref_count= 0;
+#endif /* BTR_CUR_HASH_ADAPT */
+  UT_LIST_ADD_LAST(fts_index->table->indexes, fts_index);
+}
+
+/** Append the hidden fts column and fts doc index to the
+existing table
+@param  table  table to be imported
+@param  thd    thread
+@param  cfg    metadata required by import
+@return table which has fts doc id and fts doc id index */
+static dict_table_t *build_fts_hidden_table(
+  dict_table_t *table, THD *thd, row_import &cfg)
+{
+  if (cfg.m_n_cols != static_cast<ulint>(table->n_cols) + 1
+      && cfg.m_n_indexes != UT_LIST_GET_LEN(table->indexes) + 1)
+    return nullptr;
+  dict_table_t *new_table= dict_table_t::create(
+    {table->name.m_name, strlen(table->name.m_name)},
+    table->space, table->n_t_cols - (DATA_N_SYS_COLS - 1),
+    table->n_v_cols, table->flags,
+    table->flags2);
+
+  new_table->id= table->id;
+  new_table->space_id= table->space_id;
+  size_t col_name_len= 0;
+  /* Copy columns from old table to new fts table */
+  for (ulint new_i= 0;
+       new_i < ulint(new_table->n_cols - (DATA_N_SYS_COLS + 1));
+       new_i++)
+  {
+    dict_mem_table_add_col(new_table, new_table->heap,
+                           &table->col_names[col_name_len],
+			   table->cols[new_i].mtype,
+			   table->cols[new_i].prtype,
+			   table->cols[new_i].len);
+    col_name_len+= strlen(&table->col_names[col_name_len]) + 1;
```

Comment:
Instead of maintaining an offset `col_name_len` we could maintain a pointer `col_name`:
```c++
col_name+= strlen(col_name) + 1;
```